replica/backup.go

```diff
@@ -98,19 +125,22 @@ func (rb *Backup) ReadSnapshot(snapID, volumeID string, start int64, data []byte
 
 func (rb *Backup) CloseSnapshot(snapID, volumeID string) error {
 	id := GenerateSnapshotDiskName(snapID)
-	if rb.volumeID == "" {
-		return nil
-	}
-
 	if err := rb.assertOpen(id, volumeID); err != nil {
 		return err
 	}
 
+	rb.lock.Lock()
```

Comment:
You cannot do this. assertOpen() needs to hold the lock. Between assertOpen and getting the lock again, you would have a race condition, e.g. rb.volumeID or snapshotID can change then makes the assertOpen check useless.